src/engine/CustomElements.js

```diff
@@ -0,0 +1,142 @@
+function getAttributes(file) {
+  // TODO: implement a cleaner way
+
+  const div = document.createElement("div");
+  const engine = new QmlWeb.QMLEngine(div);
+  engine.loadFile(file);
+
+  const qml = engine.rootObject;
+  return Object.keys(qml.$properties).filter(name => {
+    // Invalid names
+    if (!name.match(/^[a-z]+$/i) || name === "is") return false;
+
+    // We don't need anchors
+    if (["left", "right", "top", "bottom"].indexOf(name) !== -1) return false;
+
+    const type = qml.$properties[name].type;
+    return ["real", "color", "int", "bool", "string"].indexOf(type) !== -1;
+  });
+}
+
+function registerElement(name, file) {
+  // Delay until the document is fully loaded
+  if (document.readyState === "loading") {
+    document.addEventListener("DOMContentLoaded", () => {
+      registerElement(name, file);
+    });
+    return;
+  }
+
+  // Bail out if Custom Elements v1 are not present
+  if (!window.customElements) {
+    throw new Error(
+      "window.customElements are not supported. Consider installing a polyfill."
+    );
+  }
+
+  // We need attributes list at this point, those form a static property
+  const attributesFull = getAttributes(file);
+  const attributes = attributesFull.map(pname => pname.toLowerCase());
+  const attributesMap = attributesFull.reduce((map, pname) => {
+    map[pname.toLowerCase()] = pname;
+    return map;
+  }, {});
+
+  const QmlElement = class extends HTMLElement {
+    connectedCallback() {
+      // Default wrapper display is "block"
+      if (!this.style.display) {
+        this.style.display = "block";
+      }
+
+      const div = document.createElement("div");
+      this.appendChild(div);
+
+      // Initial height, overwritten in QML to computed height
+      // TODO: use something smarter than .offsetHeight?
+      div.style.height = `${this.offsetHeight}px`;
+
+      const engine = this.engine = new QmlWeb.QMLEngine(div);
+      engine.loadFile(file);
+      engine.start();
+      const qml = this.qml = engine.rootObject;
+
+      // Bind attributes
+      for (const attr of attributes) {
```

Comment:
Should probably avoid `for…of` for the time being, as transpliterated code for that is quite heavy and uses `Symbol.iterator`.